File: gfg-Mathematics/primeNumber.cpp
// naive approach
// #include<iostream>
// using namespace std;

// bool isPrime(int n) {
//     if(n == 1) return false;
//     for(int i = 2;i<n;i++) {
//         if(n%i == 0) return false;
//     }
//     return true;
// }

// int main() {
//     int number = 5;
//     bool find = isPrime(number);

//     if(find) cout << "The number is prime";
//     else cout <<"The number is not prime";
// }

// optimised

// #include <iostream>
// using namespace std;

// bool isPrime(int number)
// {
//     if (number == 1)
//         return false;

//     for (int i = 2; i * i <= number; i++)
//     {
//         if (number % i == 0)
//             return false;
//     }
//     return true;
// }

// int main()
// {
//     int number = 130;
//     bool find = isPrime(number);
//     if (find)
//         cout << "The number is prime";
//     else
//         cout << "The number is not prime";
// } // here we are trying to find the pair , and hence we are just going to the root of n if we find something then , the number is not prime

// more efficient method
//checking intially divisibilty for 2 and 3 

#include<iostream>
using namespace std; 

bool isPrime(int n) {
    if(n == 1) return false;
    if(n == 2 || n == 3 ) return true;

    if(n%2 == 0 || n%3 == 0) return false;

    for(int i = 5;i*i<=n;i=i+6) {
        if(n%i == 0 || n%(i+2) == 0) {
            return false;
        }
    }
    return true;
}

int main() {
    int number = 13 ;

    bool find = isPrime(number);
    if(find) cout << "The number is prime";
    else cout << "The number is not prime"; 
}

File: gfg-Mathematics/primeNumber.js
//naive
// const findPrime = (number) => {
//     if(number == 1) {
//         return false;
//     }
//     for(let i = 2;i<number;i++ ) {
//         if(number % i == 0) {
//             return false;
//         }
//     }
//     return true;
// }
// let number = 123;

// let isPrime = findPrime(number);

// if(isPrime) {
//     console.log(`${number} is a prime number`)
// }
// else {
//     console.log(`${number} is not a prime number`)
// }

//efficient method
//Going till root of n

// function isPrime(number) {
//   if (number == 1) return false;

//   for (let i = 2; i * i <= number; i++) {
//     if (number % i == 0) return false;
//   }

//   return true;
// }
// let number = 30;
// let find = isPrime(number);

// if (find) console.log("The number is prime");
// else console.log("The number is not prime number");


//more efficient method
let number = 11
const isPrime = (number) => {
    if(number == 1) return false;
    if(number == 2 || number == 3) return true;
    if(number % 2 ===0 || number % 3 == 0) return false;
    for(let i = 5;i*i<=number;i+=5){
        if(number % i == 0 || number %(i+2)==0) {
            return false;
        }
    }
    return true;
}
console.log(isPrime(number))

const findPrime = (number) => {
    if(number == 1) return false;
    if(number == 2 || number == 3) return true;
    if(number %2 == 0 || number % 3 == 0) return false;
    
    for(let i =5;i*i<=number;i+= 5) {
        if(number % i == 0) return false; 
    }
    return true;
}
console.log(findPrime(number))